src/backend/distributed/executor/intermediate_result_encoder.c

```diff
@@ -0,0 +1,324 @@
+		copyOutState->need_transcoding = (fileEncoding != databaseEncoding) ||
+										 (databaseEncodingMaxLength > 1);
+
+		encoder->copyOutState = copyOutState;
+		encoder->columnOutputFunctions =
+			ColumnOutputFunctions(tupleDesc, copyOutState->binary);
+
+		if (copyOutState->binary)
+		{
+			AppendCopyBinaryHeaders(copyOutState);
+		}
+	}
+
+	return encoder;
+}
+
+
+/*
+ * IntermediateResultEncoderReceive encodes the next row with the given encoder.
+ */
+StringInfo
+IntermediateResultEncoderReceive(IntermediateResultEncoder *encoder,
+								 Datum *values, bool *nulls)
+{
+	if (encoder->format == TEXT_COPY_FORMAT ||
+		encoder->format == BINARY_COPY_FORMAT)
+	{
+		AppendCopyRowData(values, nulls, encoder->tupleDescriptor,
+						  encoder->copyOutState, encoder->columnOutputFunctions, NULL);
+	}
+
+	if (encoder->outputBuffer->len > ENCODER_BUFFER_SIZE)
+	{
+		encoder->flushBuffer->data = encoder->outputBuffer->data;
+		encoder->flushBuffer->len = encoder->outputBuffer->len;
+
+		encoder->outputBuffer->len = 0;
+
+		return encoder->flushBuffer;
+	}
+
+	return NULL;
+}
+
+
+/*
+ * IntermediateResultEncoderDone tells the encoder that there is no more work
+ * to do. Encoder possibly can add footer data at this stage.
+ */
+StringInfo
+IntermediateResultEncoderDone(IntermediateResultEncoder *encoder)
+{
+	if (encoder->format == TEXT_COPY_FORMAT ||
+		encoder->format == BINARY_COPY_FORMAT)
+	{
+		CopyOutState copyOutState = encoder->copyOutState;
+		if (copyOutState->binary)
+		{
+			AppendCopyBinaryFooters(copyOutState);
+		}
+	}
+
+	if (encoder->outputBuffer->len > 0)
+	{
+		encoder->flushBuffer->data = encoder->outputBuffer->data;
+		encoder->flushBuffer->len = encoder->outputBuffer->len;
+
+		encoder->outputBuffer->len = 0;
+
+		return encoder->flushBuffer;
+	}
+
+	return NULL;
+}
+
+
+/*
+ * IntermediateResultEncoderDestroy cleans up resources used by the encoder.
+ */
+void
+IntermediateResultEncoderDestroy(IntermediateResultEncoder *encoder)
+{
+	if (encoder->copyOutState != NULL)
+	{
+		pfree(encoder->copyOutState);
+	}
+
+	if (encoder->columnOutputFunctions != NULL)
+	{
+		pfree(encoder->columnOutputFunctions);
+	}
+}
+
+
+/*
+ * ReadFileIntoTupleStore parses the records in the given file using the
+ * given format and puts the rows in the given tuple store.
+ */
+void
+ReadFileIntoTupleStore(char *fileName, IntermediateResultFormat format,
+					   TupleDesc tupleDescriptor, Tuplestorestate *tupstore)
+{
+	if (format == TEXT_COPY_FORMAT || format == BINARY_COPY_FORMAT)
+	{
+		char *copyFormat = (format == TEXT_COPY_FORMAT) ? "text" : "binary";
+		ReadCopyFileIntoTupleStore(fileName, copyFormat, tupleDescriptor, tupstore);
+	}
+}
+
+
+/*
+ * ReadCopyFileIntoTupleStore parses the records in a COPY-formatted file
+ * according to the given tuple descriptor and stores the records in a tuple
+ * store.
+ */
+static void
+ReadCopyFileIntoTupleStore(char *fileName, char *copyFormat,
```

Comment:
Copied from `ReadFileIntoTupleStore()` in `multi_executor.c`.

Reply:
I think we should be trying to move towards const correctness in our function signatures, especially for `char *`. `fileName` & `copyFormat` are const